docs/proposals/update-machine.md

```diff
@@ -0,0 +1,163 @@
+shortretry-->end1
+
+begin-->AddBootstrapTokenToUserData
+-->gms["statusResp,statusErr=driver.GetMachineStatus(...)"]
+-->chkstatuserr{"Checkk statusErr"}
+chkstatuserr--notFound-->chknodelbl{"Chk Node Label"}
+chkstatuserr--else-->createFailed["lastop.Type=Create,lastop.state=Failed,phase=CrashLoopBackOff|Failed"]-->medretry
+chkstatuserr--nil-->initnodename["nodeName = statusResp.NodeName"]-->setnodename
+
+
+chknodelbl--notset-->createmachine["createResp, createErr=driver.CreateMachine(...)"]-->chkCreateErr{"Check createErr"}
+
+chkCreateErr--notnil-->createFailed
+
+chkCreateErr--nil-->getnodename["nodeName = createResp.NodeName"]
+-->chkstalenode{"nodeName != machine.Name\n//chk stale node"}
+chkstalenode--false-->setnodename["if unset machine.Labels['node']= nodeName"]
+-->machinepending["if empty/crashloopbackoff lastop.type=Create,lastop.State=Processing,phase=Pending"]
+-->shortretry
+
+chkstalenode--true-->delmachine["driver.DeleteMachine(...)"]
+-->permafail["lastop.type=Create,lastop.state=Failed,Phase=Failed"]
+-->shortretry
+
+subgraph noteA [" "]
+    permafail -.- note1(["VM was referring to stale node obj"])
+end
+style noteA opacity:0
+
+
+subgraph noteB [" "]
+    createFailed-.- note2(["Proposal: Enhance handling to support startup logic"])
+end
+```
+
+### Enhancement of MCM triggerCreationFlow
+
+The following changes are proposed with a view towards minimal impact on current code and no introduction of a new Machine Phase.
+
+1. We need to introduce enhancements around the call to `Driver.CreateMachine` in the flow illustrated above.
+1. Observe that after the call to a successful `Driver.CreateMachine`, the machine phase is set to `Pending`, the `LastOperation.Type` is currently set to `Create` and the `LastOperation.State` set to `Processing`before returning with a `ShortRetry`. The `LastOperation.Description` is (unfortunately) set to the fixed message: `Creating machine on cloud provider`.
+1. Observe that after an erroneous call to `Driver.CreateMachine`, the machine phase is set to `CrashLoopBackOff` or `Failed`  (in case of creation timeout).
+1. We propose introducing a new MC error code `codes.Startup` indicating that the VM Instance was created but there was a startup error in post VM creation activities. The implementor of `Driver.CreateMachine` can return this error code, indicating that `CreateMachine` needs to be called again. However, the Machine Controller will keep the machine phase for such a situation as  `Pending` and not change the phase to `CrashLoopBackOff` when encountering a `codes.Startup` error. (The instance for the machine HAS been created hence `CrashLoopBackoff` is inappropriate)
+1. We will introduce a new _machine operation_ stage `InitiateStartup`. In case of an `codes.Startup` error, the `machine.Status.LastOperation.Description` will be set to this stage, the `LastOperation.Type` will be set to `Create` and the `LastOperation.State` set to `Failed` before returning with a `ShortRetry`
```

Comment:
I thought we said that if a `driver.CreateMachine` call fails then we simply stick to the current behavior and always move it to either `CrashLoopBackoff` or `Failed` right? `Pending` unfortunately is way too generic and today it means that `CreateMachine` call was successful but in this case it was not.

Reply:
(convo from slack)

> It was said in the meeting not to do this as `CrashLoopBackoff` is meant for VMs that don’t get created and `Failed` is meant for VMs that don’t get created after machine creation timeout.  For cases where the VM gets created but post-creation (aka startup activities) failed, we still retain the `Pending` phase

From Madhav: There was too much back and forth on this unfortunately and what i recollect is that we should retain the current behavior of moving the VM to CrashLoopBackoff because today it signifies failure in creating the VM which now involves one or more steps. Maybe @rishabh-11 and @himanshu-kun  can confirm.

Reply:
>i recollect is that we should retain the current behavior of moving the VM to CrashLoopBackoff because today it signifies failure in creating the VM.

Yes, this is correct.

> However in terms of MCM our usage is plain and simply wrong. We do not have a crashing VM

I don't think it is wrong. MCM uses `CrashloopBackoff` to say that VM creation requests are failing. if we compare it to a pod, then it is similar to pod restarts due to failure. For a VM creation, we only have success or failure, no restart happens in creation, so multiple restarts can be compared to the CreateMachine calls that MCM makes in each reconciliation until it succeeds.

Reply:
>Maybe @rishabh-11 and @himanshu-kun can confirm.

We decided to use a separate stage for the error in startup updates. Now if we want to follow the definition of CLBF, then it can be argued that startup activities should be included in it. Then in that case, we need to have proper logs that can help us differentiate why the machine is in CLBF and this should be reflected in the status of the Machine object as well.

Reply:
>I don't think it is wrong. MCM uses CrashloopBackoff to say that VM creation requests are failing. if we compare it to a pod, then it is similar to pod restarts due to failure. For a VM creation, we only have success or failure, no restart happens in creation, so multiple restarts can be compared to the CreateMachine calls that MCM makes in each reconciliation until it succeeds.

`CrashLoopBackoff` has a word crash in it for a very simple reason that the failure to start a container within a pod is that it is crashing. This means that the container gets created but it then crashes. This is not the same as failing to create a VM because the VM does not get created and then crashes.

Reply:
Then we should not use CLBF at all and replace it with something else

Reply:
>Then we should not use CLBF at all and replace it with something else

This is why i started with `It is a pity that we use Crashlookbackoff` in my original comment.

Reply:
I suggest leaving these phase renaming decisions to the overhaul. After reading the arguments above and thinking on it a bit, I agree that we can simply preserve the current phase change from `<empty>` to `CrashLoopBackoff` (yes, naming is not appropriate but this was decision taken before us) when there is an error not just on  VM instance creation but on post-creation/startup logic. 

The argument supporting this is that this is consistent phase handling on error wrt API. The argument against this is that it can confuse operators since they expect such an error code to indicate that the VM it-self was not created. However, we can mitigate this by documenting change of expected behavior in FAQ/release-notes.

Reply:
>We decided to use a separate stage for the error in startup updates. Now if we want to follow the definition of CLBF, then it can be argued that startup activities should be included in it. Then in that case, we need to have proper logs that can help us differentiate why the machine is in CLBF and this should be reflected in the status of the Machine object as well.

In the `triggerCreationFlow` when a call is made to `driver.CreateMachine`. If an error is returned then that is handled in method [machineCreateErrorHandler](https://github.com/gardener/machine-controller-manager/blob/master/pkg/util/provider/machinecontroller/machine_util.go#L480-L515) which then calls [getCreateFailurePhase](https://github.com/gardener/machine-controller-manager/blob/master/pkg/util/provider/machinecontroller/machine_util.go#L569-L582) to get the phase. With this proposal, if an update failed then an error will be returned which will have the same control flow.  The error code will be used to determine the retry interval and if the timeout has not happened then it will simply return `CrashLoopBackoff` as the phase which will then trigger invocation of `driver.CreateMachine` again.
So with minimal changes you will achieve what you wanted to, right?

Reply:
>So with minimal changes you will achieve what you wanted to, right?

yes. We just need to add the reason for CLBF in the machine status and have additional logs as well.

Reply:
> > So with minimal changes you will achieve what you wanted to, right?
> 
> yes. We just need to add the reason for CLBF in the machine status and have additional logs as well.

Will take note of this.

Reply:
Fixed